File: fpga/pynq-z2/sw/cam/sign_pre_rgb.c
/* SPDX-License-Identifier: Apache-2.0
 *
 * sign_pre_rgb.h's implementation.  Integer only, no allocation, no OS calls: it compiles
 * unchanged for the board and for the host, where B144 drives it through ctypes and checks
 * it BYTE FOR BYTE against the Python definition in modelblaster/signdet/signpre_rgb.py.
 *
 * COST.  One pass over 320x320 = 102,400 bytes with a 2-bit phase test, then 12,288 bytes of
 * white balance and quantisation.  Against the ~9.6 M cycles of the network it is noise --
 * the same argument sign_pre.c makes for the grey path, and it should be measured on the
 * board rather than believed.
 */
#include "sign_pre_rgb.h"

void sign_pre_rgb64(const uint8_t *frame, uint8_t *rgb64)
{
	int oy, ox, dy, dx;

	for (oy = 0; oy < SPR_OUT; oy++) {
		for (ox = 0; ox < SPR_OUT; ox++) {
			uint32_t sum[3] = {0, 0, 0};
			uint32_t cnt[3] = {0, 0, 0};
			int ry, rx, c;

			for (dy = 0; dy < SPR_BOX; dy++) {
				/* absolute row/col within the 320 crop -- the parity here IS
				 * the CFA phase, and it is why the mask cannot be block-local. */
				int i = oy * SPR_BOX + dy;
				const uint8_t *p = frame
					+ (size_t)(SPR_OFF + i) * SPR_STRIDE
					+ SPR_PAD + SPR_OFF + ox * SPR_BOX;

				for (dx = 0; dx < SPR_BOX; dx++) {
					int j = ox * SPR_BOX + dx;
					/* BGGR: (even,even)=B  (odd,odd)=R  else G */
					int ch = ((i & 1) && (j & 1)) ? 0
					       : (!(i & 1) && !(j & 1)) ? 2 : 1;

					sum[ch] += p[dx];
					cnt[ch] += 1;
				}
			}
			/* round(sum/cnt) = (2*sum + cnt) / (2*cnt), integer. */
			ry = SPR_OUT - 1 - oy;          /* the 180 rotation, applied to the */
			rx = SPR_OUT - 1 - ox;          /* finished RGB, never to the mosaic */
			for (c = 0; c < 3; c++) {
				rgb64[(ry * SPR_OUT + rx) * 3 + c] =
					(uint8_t)((2u * sum[c] + cnt[c]) / (2u * cnt[c]));
			}
		}
	}
}

void sign_pre_rgb_wb(uint8_t *rgb64)
{
	int c, i;

	for (c = 0; c < 3; c++) {
		uint32_t s = 0, m;

		for (i = 0; i < SPR_PIX; i++) {
			s += rgb64[i * 3 + c];
		}
		m = s / SPR_PIX;                        /* floor, matching the Python */
		if (m < 1u) { m = 1u; }
		for (i = 0; i < SPR_PIX; i++) {
			uint32_t v = ((uint32_t)rgb64[i * 3 + c] * SPR_WB_TARGET + m / 2u) / m;

			rgb64[i * 3 + c] = (uint8_t)(v > 255u ? 255u : v);
		}
	}
}

void sign_pre_rgb_quant(const uint8_t *rgb64, int8_t *out)
{
	int c, i;

	/* HWC -> CHW: conv1 reads NCHW, so the three planes must be contiguous. */
	for (c = 0; c < 3; c++) {
		for (i = 0; i < SPR_PIX; i++) {
			uint32_t q = (2u * (uint32_t)SPR_IN_SCALE_RECIP
				      * (uint32_t)rgb64[i * 3 + c] + 255u) / 510u;

			if (q > 127u) { q = 127u; }
			out[c * SPR_PIX + i] = (int8_t)q;
		}
	}
}

int sign_pre_rgb_selftest(uint8_t *st_frame)
{
	uint8_t rgb[SPR_ELEMS];
	int i, j;

	/* A frame that is constant in each CFA plane must come back as that exact colour --
	 * this is the check that catches a phase error, because R and B are different here. */
	for (i = 0; i < SPR_H; i++) {
		for (j = 0; j < SPR_W; j++) {
			int ii = i - SPR_OFF, jj = j - SPR_OFF;
			uint8_t v = ((ii & 1) && (jj & 1)) ? 200      /* R */
				  : (!(ii & 1) && !(jj & 1)) ? 40     /* B */
				  : 120;                              /* G */

			st_frame[i * SPR_STRIDE + SPR_PAD + j] = v;
		}
	}
	sign_pre_rgb64(st_frame, rgb);
	for (i = 0; i < SPR_PIX; i++) {
		if (rgb[i * 3 + 0] != 200) { return 1; }
		if (rgb[i * 3 + 1] != 120) { return 2; }
		if (rgb[i * 3 + 2] != 40)  { return 3; }
	}
	return 0;
}

File: fpga/pynq-z2/modelblaster/signdet/signpre_rgb.py
"""Lab B144 colour front end: HM01B0 Bayer frame -> 64x64x3 RGB, integer-exact.

THE DEPLOY CONTRACT.  Everything the detector is trained on goes through this function, and
the board-side C (sign_pre_rgb64) must reproduce it byte for byte.  Kept in Python here so
the training set, the host harness and the C can all be checked against one definition.

WHY NOT DEMOSAIC.  sign_pre.c's grey path averages RAW Bayer bytes over an EVEN box, because
any even box holds equal numbers of R, G and B and so needs no CFA phase.  A colour front end
cannot dodge the phase that way -- it must commit.  But it does not need a demosaic either:
for each 5x5 output block we accumulate each CFA colour into its OWN accumulator and divide by
its OWN count.  That is an unbiased per-channel mean over the 4..9 samples of that colour in
the block, it needs no interpolation, and it is what kills the speckle:

  measured on snap_015, the bilinear demosaic of cam_view.py leaves a red-excess signal of
  ~15/255 on the sign under speckle whose peak amplitude is 171 -- the noise is 10x the
  signal.  Averaging 4..9 same-colour samples per block removes the speckle by construction,
  because the speckle IS the interpolation error and we never interpolate.

THE PHASE IS BGGR, and it is not a guess: under BGGR the frame's illuminant reads warm
(R~=G>B, which is the fluorescent light you can see in the picture) and the props read red.
Under RGGB the same frame reads cyan and the STOP sign reads blue.

GEOMETRY.  324x324 with a 326-byte stride and 2 pad bytes per line.  We take rows/cols 2..321
-- a 320x320 crop, offset 2, EVEN, so the CFA phase is still BGGR -- and tile it with 5x5
blocks to 64x64.  The 2-pixel border we drop is vignetted anyway.  THIS IS NOT THE CENTRE CROP
THE LAB RULED OUT: it discards 2 pixels, not the 62% of the frame a sign-sized crop would.
The 180 rotation is applied to the 64x64x3 RGB, AFTER the mosaic is read, so it can never
disturb the phase.
"""
import numpy as np

STRIDE, W, H, PAD = 326, 324, 324, 2
CROP, OUT, BOX = 320, 64, 5
OFF = (W - CROP) // 2          # 2, even -> CFA phase preserved
WB_TARGET = 110                # gray-world target mean per channel


def read_raw(path):
    a = np.fromfile(path, dtype=np.uint8)
    if a.size != STRIDE * H:
        raise ValueError("%s: %d bytes, expected %d" % (path, a.size, STRIDE * H))
    return a.reshape(H, STRIDE)[:, PAD:PAD + W]


def bayer_to_rgb64(bayer324, rot180=True):
    """320x320 BGGR mosaic -> 64x64x3 uint8 by per-channel counted 5x5 box mean.

    THE BOX IS ODD (5) AND THAT MATTERS.  The CFA phase of a pixel is the parity of its
    ABSOLUTE row/col, and with an odd box the phase of a block's first pixel alternates with
    the block index.  A mask fixed in block-local coordinates therefore averages R and B
    together on half the blocks and returns a grey image -- measured, on the bench frames it
    collapsed the sign to R82.6 G83.2 B83.4.  So the masks below are built over the WHOLE
    crop from absolute coordinates, and each block divides by the count it actually has
    (R and B are 4, 6 or 9 per block; G is 8, 12 or 13).  The crop offset is even, so
    in-crop parity is frame parity.
    """
    b = bayer324[OFF:OFF + CROP, OFF:OFF + CROP].astype(np.uint32)
    yy, xx = np.meshgrid(np.arange(CROP), np.arange(CROP), indexing="ij")
    sel = {"R": (yy % 2 == 1) & (xx % 2 == 1),
           "B": (yy % 2 == 0) & (xx % 2 == 0),
           "G": (yy % 2) != (xx % 2)}
    planes = {}
    for name, m in sel.items():
        s = (b * m).reshape(OUT, BOX, OUT, BOX).sum(axis=(1, 3))
        c = m.astype(np.uint32).reshape(OUT, BOX, OUT, BOX).sum(axis=(1, 3))
        planes[name] = ((2 * s + c) // (2 * c)).astype(np.uint8)   # round-half-up
    rgb = np.dstack([planes["R"], planes["G"], planes["B"]])
    return np.rot90(rgb, 2).copy() if rot180 else rgb


def gray_world(rgb64):
    """Per-channel gain to a common mean. Integer, and it is what removes the illuminant."""
    out = np.empty_like(rgb64)
    for c in range(3):
        ch = rgb64[:, :, c].astype(np.uint32)
        m = int(ch.mean())
        if m < 1:
            m = 1
        out[:, :, c] = np.minimum((ch * WB_TARGET + m // 2) // m, 255).astype(np.uint8)
    return out


def preprocess(path_or_bayer, wb=True):
    b = read_raw(path_or_bayer) if isinstance(path_or_bayer, str) else path_or_bayer
    rgb = bayer_to_rgb64(b)
    return gray_world(rgb) if wb else rgb


def mosaic_bggr(rgb324):
    """Inverse of the read: an RGB scene -> the single-byte-per-pixel BGGR frame a sensor
    would deliver. Used to put training images through the same optics as the board."""
    h, w = rgb324.shape[:2]
    yy, xx = np.meshgrid(np.arange(h), np.arange(w), indexing="ij")
    out = np.where((yy % 2 == 1) & (xx % 2 == 1), rgb324[:, :, 0],
          np.where((yy % 2 == 0) & (xx % 2 == 0), rgb324[:, :, 2], rgb324[:, :, 1]))
    return out.astype(np.uint8)
